File: templates/python/code-carl/ch11_graph/t1_lc0797_all_paths_source_target.py
"""LeetCode 797. All Paths From Source to Target
https://leetcode.cn/problems/all-paths-from-source-to-target/
"""

from typing import List


class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        pass

    def dfs(self, curr: int, end: int) -> None:
        pass


if __name__ == "__main__":
    solution = Solution()

    result = sorted(solution.allPathsSourceTarget([[1, 2], [3], [3], []]))
    assert result == sorted([[0, 1, 3], [0, 2, 3]])

    result = sorted(solution.allPathsSourceTarget([[4, 3, 1], [3, 2, 4], [3], [4], []]))
    assert result == sorted([[0, 4], [0, 3, 4], [0, 1, 3, 4], [0, 1, 2, 3, 4], [0, 1, 4]])

    print("All tests passed.")

File: templates/code-carl/ch11_graph/T1_LC0797_AllPathsSourceTarget.java
/**
 * Practice template for:
 * - LeetCode 797. All Paths From Source to Target
 *   https://leetcode.cn/problems/all-paths-from-source-to-target/
 */
import java.util.ArrayList;
import java.util.List;

/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/12/2 09:21
 * @comment
 */
public class T1_LC0797_AllPathsSourceTarget {
    List<Integer> path = new ArrayList<>();
    List<List<Integer>> result = new ArrayList<>();
    int[][] graph;

    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        int n = graph.length - 1;
        this.graph = graph;
        path.add(0);
        dfs(0, n);
        return result;
    }

    public void dfs(int curr, int end) {
        if (curr == end) {
            result.add(new ArrayList<>(path));
        }
        for (int next : graph[curr]) {
            path.add(next);
            dfs(next, end);
            path.remove(path.size() - 1);
        }
    }
}
